internal/cache/file/downloader/job.go

```diff
@@ -0,0 +1,141 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package downloader
+
+import (
+	"container/list"
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/locker"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	"golang.org/x/net/context"
+)
+
+type jobStatusName string
+
+const (
+	NOT_STARTED jobStatusName = "NOT_STARTED"
+	DOWNLOADING jobStatusName = "DOWNLOADING"
+	COMPLETED   jobStatusName = "COMPLETED"
+	FAILED      jobStatusName = "FAILED"
+	CANCELLED   jobStatusName = "CANCELLED"
+)
+
+// Job downloads the requested object from GCS into the specified local file
+// path with given permissions and ownership.
+type Job struct {
+	/////////////////////////
+	// Constant data
+	/////////////////////////
+
+	object               *gcs.MinObject
+	bucket               gcs.Bucket
+	filePath             string
+	fileInfoCache        *lru.Cache
+	sequentialReadSizeMb int32
+	filePerm             os.FileMode
+	fileUid              uint32
+	fileGid              uint32
+
+	/////////////////////////
+	// Mutable state
+	/////////////////////////
+
+	// Represents the current status of Job.
+	status JobStatus
+
+	// list of subscribers waiting on async download.
+	//
+	// INVARIANT: Each element is of type jobSubscriber
+	subscribers list.List
+
+	// Context & its CancelFunc for cancelling async download in progress.
+	cancelCtx  context.Context
+	cancelFunc context.CancelFunc
+
+	mu locker.Locker
+}
+
+// JobStatus represents the status of job.
+type JobStatus struct {
+	Name   jobStatusName
+	Err    error
+	Offset int64
+}
+
+// jobSubscriber represents a subscriber waiting on async download of job to
+// complete downloading at least till the subscribed offset.
+type jobSubscriber struct {
+}
+
+func NewJob(object *gcs.MinObject, bucket gcs.Bucket, filePath string,
+	fileInfoCache *lru.Cache, sequentialReadSizeMb int32, filePerm os.FileMode,
+	fileUid uint32, fileGid uint32) (job *Job) {
+	job = &Job{
+		object:               object,
+		bucket:               bucket,
+		filePath:             filePath,
+		fileInfoCache:        fileInfoCache,
+		sequentialReadSizeMb: sequentialReadSizeMb,
+		filePerm:             filePerm,
+		fileUid:              fileUid,
+		fileGid:              fileGid,
+	}
+	job.mu = locker.New("Job-"+filePath, job.checkInvariants)
+	job.init()
+	return
+}
+
+// CheckInvariants panic if any internal invariants have been violated.
+func (job *Job) checkInvariants() {
+	// INVARIANT: Each subscriber is of type jobSubscriber
+	for e := job.subscribers.Front(); e != nil; e = e.Next() {
+		switch e.Value.(type) {
+		case jobSubscriber:
```

Comment:
Why switch case? - We will not support types other than jobSubscriber?

Reply:
I couldn't find a simple/efficient way to do this using if statement. So, will continue with switch.

Reply:
Tried this - https://go.dev/play/p/0QP63lALIt4